src/pudl/io_managers.py

```diff
@@ -0,0 +1,592 @@
+
+        Read about the PRAGMAs here: https://www.sqlite.org/pragma.html#pragma_foreign_key_check
+
+        Raises:
+            ForeignKeyErrors: if data in the database violate foreign key constraints.
+        """
+        logger.info(f"Running foreign key check on {self.db_name} database.")
+        with self.engine.connect() as con:
+            fk_errors = pd.read_sql_query("PRAGMA foreign_key_check;", con)
+
+        if not fk_errors.empty:
+            # Merge in the actual FK descriptions
+            tables_with_fk_errors = fk_errors.table.unique().tolist()
+            table_foreign_keys = pd.concat(
+                [self._get_fk_list(table) for table in tables_with_fk_errors]
+            )
+
+            fk_errors_with_keys = fk_errors.merge(
+                table_foreign_keys,
+                how="left",
+                on=["parent", "fkid", "table"],
+                validate="m:1",
+            )
+
+            errors = []
+            # For each foreign key error, raise a ForeignKeyError
+            for (
+                table_name,
+                parent_name,
+                parent_fk,
+            ), parent_fk_df in fk_errors_with_keys.groupby(["table", "parent", "fk"]):
+                errors.append(
+                    ForeignKeyError(
+                        child_table=table_name,
+                        parent_table=parent_name,
+                        foreign_key=parent_fk,
+                        rowids=parent_fk_df["rowid"].values,
+                    )
+                )
+            raise ForeignKeyErrors(errors)
+
+    def _handle_pandas_output(self, context: OutputContext, df: pd.DataFrame):
+        """Write dataframe to the database.
+
+        Args:
+            context: dagster keyword that provides access output information like asset name.
+            df: dataframe to write to the database.
+        """
+        table_name = self._get_table_name(context)
+
+        sa_table = self._get_sqlalchemy_table(table_name)
+        engine = self.engine
+
+        # TODO (bendnorman): I included this if else statement for the analysis table example.
+        if sa_table is None:
+            with engine.connect() as con:
+                # Remove old table records before loading to db
+                df.to_sql(
+                    table_name,
+                    con,
+                    if_exists="replace",
+                    index=False,
+                )
+        else:
+            with engine.connect() as con:
+                # Remove old table records before loading to db
+                con.execute(sa_table.delete())
+
+                df.to_sql(
+                    table_name,
+                    con,
+                    if_exists="append",
+                    index=False,
+                    dtype={c.name: c.type for c in sa_table.columns},
+                )
+
+    # TODO (bendnorman): Create a SQLQuery type so it's clearer what this method expects
```

Comment:
also out-of-scope/food-for-thought: Could we maybe use SQLAlchemy's types here? After about 30s skimming this [views recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/Views) [`Selectable`](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.Selectable) might work...

Reply:
That sounds like a good idea! I'll have to look into it.